src/particles.zig

```diff
@@ -0,0 +1,441 @@
+			isEmissionBroken = true;
+		}
+
+		if(isBaseBroken) {
+			createDummyAnimationFrames(&textures, baseAnimationFrameCount, Image.defaultImage);
+			createDummyAnimationFrames(&emissionTextures, baseAnimationFrameCount, Image.emptyImage);
+		} else if(isEmissionBroken or !hasEmission) {
+			createAnimationFrames(&textures, baseAnimationFrameCount, base);
+			createDummyAnimationFrames(&emissionTextures, baseAnimationFrameCount, Image.emptyImage);
+		} else {
+			createAnimationFrames(&textures, baseAnimationFrameCount, base);
+			createAnimationFrames(&emissionTextures, baseAnimationFrameCount, emission);
+		}
+
+		return typ;
+	}
+
+	fn readTexture(assetsFolder: []const u8, textureId: []const u8, suffix: []const u8, default: graphics.Image, status: enum {isOptional, isMandatory}) graphics.Image {
+		var splitter = std.mem.splitScalar(u8, textureId, ':');
+		const mod = splitter.first();
+		const id = splitter.rest();
+
+		const gameAssetsPath = std.fmt.allocPrint(main.stackAllocator.allocator, "assets/{s}/particles/textures/{s}{s}", .{mod, id, suffix}) catch unreachable;
+		defer main.stackAllocator.free(gameAssetsPath);
+
+		const worldAssetsPath = std.fmt.allocPrint(main.stackAllocator.allocator, "{s}/{s}/particles/textures/{s}{s}", .{assetsFolder, mod, id, suffix}) catch unreachable;
+		defer main.stackAllocator.free(worldAssetsPath);
+
+		return graphics.Image.readFromFile(arenaAllocator, worldAssetsPath) catch graphics.Image.readFromFile(arenaAllocator, gameAssetsPath) catch {
+			if(status == .isMandatory) std.log.err("Particle texture not found in {s} and {s}.", .{worldAssetsPath, gameAssetsPath});
+			return default;
+		};
+	}
+
+	fn createDummyAnimationFrames(container: *main.List(Image), frameCount: usize, image: Image) void {
+		for(0..frameCount) |_| {
+			container.append(image);
+		}
+	}
+
+	fn createAnimationFrames(container: *main.List(Image), frameCount: usize, image: Image) void {
+		for(0..frameCount) |i| {
+			container.append(extractAnimationSlice(image, i));
+		}
+	}
+
+	fn extractAnimationSlice(image: Image, fameIndex: usize) Image {
+		const frameCount = image.height/image.width;
+		const frameHeight = image.height/frameCount;
+		const startHeight = frameHeight*fameIndex;
+		const endHeight = frameHeight*(fameIndex + 1);
+		var result = image;
+		result.height = @intCast(frameHeight);
+		result.imageData = result.imageData[startHeight*image.width .. endHeight*image.width];
+		return result;
+	}
+
+	pub fn generateTextureArray() void {
+		textureArray.generate(textures.items, true, true);
+		c.glTexParameterf(c.GL_TEXTURE_2D_ARRAY, c.GL_TEXTURE_MAX_ANISOTROPY, @floatFromInt(main.settings.anisotropicFiltering));
+		emissionTextureArray.generate(emissionTextures.items, true, false);
+		c.glTexParameterf(c.GL_TEXTURE_2D_ARRAY, c.GL_TEXTURE_MAX_ANISOTROPY, @floatFromInt(main.settings.anisotropicFiltering));
+
+		particleTypesSSBO = SSBO.initStatic(ParticleType, ParticleManager.types.items);
```

Comment:
Please init it in the init function to avoid creating it twice when the texture array is regenerated.
Since you don't know the initial size, you can just use `particleTypesSSBO.bufferData` here.